### PR TITLE
[ML] Add prefix strings option to trained models

### PR DESCRIPTION
**Pull request by davidkyle**
Certain NLP models such as [multilingual-e5-large](https://huggingface.co/intfloat/multilingual-e5-large) require a prefix string to be applied to the input text. For asymmetric tasks such as information retrieval the prefix can be different when ingesting the data and when searching it. For example text embedding model can have a one prefix applied when the model is evaluated as part of an knn search and a different prefix when ingesting documents. 

An example model configuration with prefix strings is:
```
{
    "model_type": "pytorch",
    "inference_config": {
        "text_embedding": {
            "tokenization": {
                "xlm_roberta": {
                    "do_lower_case": false,
                    "with_special_tokens": true,
                    "max_sequence_length": 512,
                    "truncate": "first",
                    "span": -1
                }
            },
            "embedding_size": 384
        }
    },
    "prefix_strings": {
        "search": "this is a query",
        "ingest": "this is a passage"
    }
}
```

Many files have been touched by this change but the bulk of the work is quite simple: define the configuration object an pass the prefix type (context) parameter down the inference calls.

### CODE REVIEW

File: docs/reference/ml/trained-models/apis/put-trained-models.asciidoc [hunk 1]
```diff
@@ -701,6 +701,32 @@ the platform identifiers used by Elasticsearch, so one of, `linux-x86_64`,
...
+in the search context and one that is always applied when ingesting the
+docs. Both are optional.
++
+.Properties of `prefix_strings`
+[%collapsible%open]
+====
+`search`:::
+(Optional, string)
+The prefix string to prepend to the input text for requests
+originating from a search query.
```
Comment (szabosteve) [inline docs/reference/ml/trained-models/apis/put-trained-models.asciidoc:723]:
```suggestion
originating from a search query.

```

File: docs/reference/ml/trained-models/apis/put-trained-models.asciidoc [hunk 2]
```diff
@@ -701,6 +701,32 @@ the platform identifiers used by Elasticsearch, so one of, `linux-x86_64`,
...
+`ingest`:::
+(Optional, string)
+The prefix string to prepend to the input text for requests
+at ingest where the Inference ingest processor is used. // TODO is there a shortcut for Inference ingest processor?
```
Comment (szabosteve) [inline docs/reference/ml/trained-models/apis/put-trained-models.asciidoc:727]:
```suggestion
at ingest where the {infer} ingest processor is used. // TODO is there a shortcut for Inference ingest processor?
```

Comment (szabosteve) [reply docs/reference/ml/trained-models/apis/put-trained-models.asciidoc:727]: There is no attribute to shorten `inference ingest processor`. Do you want me to create one? Or `{infer}` for inference is enough?

**Review by droberts195 — APPROVED** [review]
LGTM, but if we're going to have packaged models on GCS that use this feature then the new section will need to also go in this class: https://github.com/elastic/elasticsearch/blob/8d6ded3736ec12dc56d2edd6f0d493916debbc45/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ModelPackageConfig.java#L33

**Review by szabosteve — APPROVED** [review]
Docs LGTM, thanks!

**Comment by davidkyle** [comment]
@elasticmachine update branch

**Comment by davidkyle** [comment]
Closing as I've made a mess with git. #102089 is raised as a replacement

**Pull request closed without merging**